odl/operator/pspace_ops.py

```diff
@@ -0,0 +1,190 @@
+﻿# Copyright 2014, 2015 Jonas Adler
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Default operators defined on any (reasonable) space."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+import scipy as sp
+
+# ODL imports
+from odl.operator.operator import Operator
+from odl.set.pspace import ProductSpace
+
+
+__all__ = ('ProductSpaceOperator', 'ComponentProjection')
+
+
+class ProductSpaceOperator(Operator):
+    """A separable operator on product spaces."""
+
+    def __init__(self, operators, dom=None, ran=None):
+        """ TODO
+        """
+                   
+        # Validate input data
+        if dom is not None and not isinstance(dom, ProductSpace):
+            raise TypeError('space {!r} not a ProductSpace instance.'
+                            ''.format(dom))
+        if ran is not None and not isinstance(ran, ProductSpace):
+            raise TypeError('space {!r} not a ProductSpace instance.'
+                            ''.format(dom))
+
+        # Convert ops to sparse representation
+        self.ops = sp.sparse.coo_matrix(operators)
```

Comment:
I fiddled around a bit with sparse matrix, and it seems impossible to initialize a CSR matrix without running into some C library code. It's simply not made for `dtype==object`, and when one creates an empty matrix with that type, there is an explicit warning about this.

That does not stop us from trying to use these operators a bit like matrices, especially when doing arithmetic. With dense matrices, it works out of the box, in fact, so for that part I guess (unless we're talking about huge operator arrays) we can simply use that, while for evaluation we store them in a sparse matrix as now. How about that?